### ISSUE TITLE
Need for a Precompile makefile

### ISSUE THREAD

**Issue by rasmus-madsen** [lowRISC/opentitan#1377]
I think we need a more flexible build system for the IP parts that has custom setup needs
like the AES module. at least until FuseSoc reaches a state where such things can be added there.
the AES dv need to pass a customized input to the compiler to get it to recognize the c-files.
this input is dependent on the compiler.

Currently we are setting up the choice of compiler inside the 
`hw/dv/tools/Makefile`

```
\# Make GCC find aes.h
ifeq (${SIMULATOR},vcs)
  BUILD_OPTS += -CFLAGS "-I${DV_DIR}/../model"
else ifeq(${SIMULATOR},xcelium)
  BUILD_OPTS += -I${DV_DIR}/../model
endif

\# Make GCC link OpenSSL/BoringSSL
BUILD_OPTS += -lcrypto
```
unfortunately this enforce either exporting the SIMULATOR locally or setting it from command line everytime.

I suggest we add a `pre_comp.mk` or similar to setup build related environment variables prior to running the local `dv/Makefile`. this will make the tool flow more flexible

this should setup the `SIMULATOR ?= vcs`
I suggest putting other general parameters like
```
MAKE_ROOT
SHELL
SEED
TIMESTAMP
```
to make them accessible in the local `dv/Makefile`.

unfortunately this will require some updating of what is already in place and the autogenerate tool from Sri

**Comment by rasmus-madsen** [comment]
hmm the codeblocks dissappeard. I hope you can read it.

**Comment by vogelpi** [comment]
I think having a `pre_comp.mk` as suggested by @rasmus-madsen is a good idea. It would ease the use of our DV infrastructure for designers, e.g., for debugging.

**Comment by weicaiyang** [comment]
@rasmus-madsen 
We're migrating to another sim/regression script. Perhaps we should find a way to set simulator dependent option for the new script. For current makefile flow, maybe just use a simple workaround since it's about to be deprecated.
@sriyerg any thought?

**Comment by sriyerg** [comment]
This is one of the many quirks of using Makefiles - even though the default for SIMULATOR is set to vcs in `hw/dv/tools/Makefile`, constructs like these don't work. There is a file order dependency when including sub-makefiles. I think adding the simulator-dependent build option at the end of `aes/dv/Makefile` (after the line that includes the common Makefile) would probably work. 

This ambiguity is one of the reasons why we want to move away from using Makefiles. We are building up our regression tool in `util/dvsim.py` and support for this construct can be added in a much cleaner way. A fix for it within the new tool is now available in PR #1385.

**Comment by rasmus-madsen** [comment]
@sriyerg not sure this is a quirk ?
it makes sense that the parameter has to be set before you can check it.
your suggestion will probably not work since the rules.mk is called inside tools/Makefile which would out our option after the rules.
for a make based flow I think it makes very good sense to have a pre compile file that setup environment related things before anything else happens. 
anyway if we are moving away from Make I think the current solutions is ok until then.

**Comment by sriyerg** [comment]
I just tried it and it worked for me, which is why it is a little quirky. This means the value of $(SIMULATOR) is undefined when it encounters the conditional statement (because it happens during the initial parsing step); whereas it is 'vcs' when it executes the target recipes. Using conditionals to set other variables requires an order to be followed, but target and recipes can be placed anywhere. Not everyone has a compete understanding of how all this works, and it leads to some debug overhead. 

The `util/dvsim.py` is minimally viable at this point - it still needs addition of several features and fixes. It is already a superset of the Make based flow and supports VCS and Xcelium. I'd suggest using it to run tests going forward. I still have to add documentation for it, but the `--help` switch should be of help for now.

**Comment by rasmus-madsen** [comment]
I tihnk we can avoid some confusion by changing the code to this

```
\# Make GCC find aes.h
ifeq(${SIMULATOR},xcelium)
  BUILD_OPTS += -I${DV_DIR}/../model
else
  BUILD_OPTS += -CFLAGS "-I${DV_DIR}/../model"
endif

\# Make GCC link OpenSSL/BoringSSL
BUILD_OPTS += -lcrypto
```

**Comment by sriyerg** [comment]
Sure, this works too.

**Comment by vogelpi** [comment]
The suggestion of @rasmus-madsen for the Makefile flow got merged in #1405 and we are moving away from using Makefiles in the future. I think we can close this one, but feel free to re-open if you do not agree.
